### PR TITLE
Add gitIngestion flag to use FileIngestion::Git

### PR DESCRIPTION
**Pull request by matthewbauer**
This allows you to omit the treeHash value.

### CODE REVIEW

**Comment by Ericson2314** [comment]
Let's do #33 instead. Submodules don't work for git hashes, remember.

**Comment by matthewbauer** [comment]
There's nothing stopping you from using git ingestion with submodules - the hash just won't match what's in the git repo

File: src/libfetchers/git.cc [hunk 1]
```diff
@@ -104,7 +109,8 @@ struct GitInput : Input
         assert(!rev || rev->type == htSHA1);
         assert(!treeHash || treeHash->type == htSHA1);
 
-        auto ingestionMethod = treeHash ? FileIngestionMethod::Git : FileIngestionMethod::Recursive;
+        if (submodules && treeHash)
```
Comment (Ericson2314) [inline src/libfetchers/git.cc:112]:
```suggestion
        if (submodules && ingestionMethod == FileIngestionMethod::Git)
```

Comment (matthewbauer) [reply src/libfetchers/git.cc:112]: i don't think that's necessary - ingestionMethod can have git submodules, they just will be different from the tree hash

File: src/libfetchers/git.cc [hunk 2]
```diff
@@ -501,6 +507,9 @@ struct GitInputScheme : InputScheme
 
         input->submodules = maybeGetBoolAttr(attrs, "submodules").value_or(false);
 
+        if (maybeGetBoolAttr(attrs, "gitIngestion").value_or(false) || input->treeHash)
```
Comment (Ericson2314) [inline src/libfetchers/git.cc:510]:
```suggestion
        if (maybeGetBoolAttr(attrs, "gitIngestion").value_or(input->treeHash))
```
or if that doesn't work
```suggestion
        if (maybeGetBoolAttr(attrs, "gitIngestion").value_or((bool) input->treeHash))
```

File: src/libfetchers/git.cc [hunk 3]
```diff
@@ -256,9 +262,12 @@ struct GitInput : Input
 
         if (isLocal) {
 
-            if (!input->rev && !input->treeHash)
+            if (!input->rev)
                 input->rev = Hash(chomp(runProgram("git", true, { "-C", actualUrl, "rev-parse", *input->ref })), htSHA1);
 
+            if (!input->treeHash)
```
Comment (Ericson2314) [inline src/libfetchers/git.cc:268]:
```suggestion
            if (!submodules && !input->treeHash)
```

File: src/libfetchers/git.cc (hunk 1 above)
Comment (Ericson2314) [reply src/libfetchers/git.cc:112]: Yes good point, maybe add warning in that case?

File: src/libfetchers/git.cc [hunk 4]
```diff
@@ -104,7 +109,8 @@ struct GitInput : Input
         assert(!rev || rev->type == htSHA1);
         assert(!treeHash || treeHash->type == htSHA1);
 
-        auto ingestionMethod = treeHash ? FileIngestionMethod::Git : FileIngestionMethod::Recursive;
+        if (submodules && treeHash)
+            throw Error("Cannot combine tree hashes with git submodules");
```
Comment (Ericson2314) [inline src/libfetchers/git.cc:113]:
```suggestion
        if (submodules) {
            if (treeHash)
                throw Error("Cannot combine tree hashes with git submodules");
            warn("git tree hash will be different when submodules are converted to regular directories");
```

**Pull request merged**